shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCXConnection.scala

```diff
@@ -33,129 +34,132 @@ import org.apache.spark.internal.Logging
     ucx.startManagementPort(host)
   }
 
-  override def send(sendPeerExecutorId: Long, buffer: AddressLengthTag,
-    cb: TransactionCallback): Transaction =
-    send(sendPeerExecutorId, buffer, Seq.empty, cb)
-
-  override def registerRequestHandler(
-      requestType: RequestType.Value, cb: TransactionCallback): Unit = {
-
+  override def registerRequestHandler(messageType: MessageType.Value,
+      cb: TransactionCallback): Unit = {
     ucx.registerRequestHandler(
-        UCXConnection.composeRequestAmId(requestType), () => new UCXAmCallback {
-      private val tx = createTransaction
-
-      tx.start(UCXTransactionType.Request, 1, cb)
-
-      override def onSuccess(am: UCXActiveMessage, buff: RefCountedDirectByteBuffer): Unit = {
-        logDebug(s"At requestHandler for ${requestType} and am: " +
-          s"$am")
-        tx.completeWithSuccess(requestType, Option(am.header), Option(buff))
-      }
+      UCXConnection.composeRequestAmId(messageType), () =>
+        new UCXAmCallback {
+          private val tx = createTransaction
+          tx.start(UCXTransactionType.Request, 1, cb)
+
+          override def onSuccess(am: UCXActiveMessage, buff: TransportBuffer): Unit = {
+            logDebug(s"At requestHandler for ${messageType} and am: " +
+              s"$am")
+            tx.completeWithSuccess(messageType, Option(am.header), Option(buff))
+          }
+
+          override def onMessageReceived(size: Long, header: Long,
+                                         finalizeCb: (TransportBuffer => Unit)): Unit = {
+            logDebug(s"onMessageReceived for hdr ${TransportUtils.toHex(header)}")
+            finalizeCb(new MetadataTransportBuffer(transport.getDirectByteBuffer(size)))
+          }
+
+          override def onError(am: UCXActiveMessage, ucsStatus: Int, errorMsg: String): Unit = {
+            tx.completeWithError(errorMsg)
+          }
+
+          override def onCancel(am: UCXActiveMessage): Unit = {
+            tx.completeCancelled(messageType, am.header)
+          }
+
+          override def onMessageStarted(receiveAm: UcpRequest): Unit = {
+            tx.registerPendingMessage(receiveAm)
+          }
+        })
+  }
 
-      override def onHostMessageReceived(size: Long): RefCountedDirectByteBuffer = {
-        transport.getDirectByteBuffer(size)
-      }
+  override def send(peerExecutorId: Long,
+      messageType: MessageType.Value,
+      header: Long,
+      buffer: MemoryBuffer,
+      cb: TransactionCallback): Transaction = {
+    val tx = createTransaction
+    tx.start(UCXTransactionType.Send, 1, cb)
+    logDebug(s"Sending to ${peerExecutorId} at ${TransportUtils.toHex(header)} " +
+      s"with ${buffer}")
 
-      override def onError(am: UCXActiveMessage, ucsStatus: Int, errorMsg: String): Unit = {
-        tx.completeWithError(errorMsg)
-      }
+    val sendAm = UCXActiveMessage(
+      UCXConnection.composeSendAmId(messageType), header,
+      buffer.isInstanceOf[DeviceMemoryBuffer])
```

Comment:
This is actually a bug. The third argument is `forceRndv` we want rndv for host buffers in addition for gpu buffers, but in the future we may want UCX to do this, otherwise we may end up in an eager mode, which means we need to copy in the progress thread in a blocking fashion.